Generate synthetic transcripts from commits

jakethekoenig opened the pull request:
A script that:
* Iterates over github commits
* Uses GPT to make a request and step by step summary
* Converts the diff to a replacement format llm response
* packages them into training examples

Usage:
```
./scripts/git_log_to_transcript.py
```
will iterate over the last 10 commits in mentat and convert them. There are flags to choose the repo, the start commit, and the number of commits.

This is very close to being able to generate code context benchmarks similar to @granawkins's. I plan to do that in a followup PR.

PCSwingle reviewed (approved):
I love this! How good is gpt at creating the user requests and the model conversation outputs?

biobootloader commented:
Awesome! Can't wait to see what fine tuned models from this data do.

Can you post an example or two of the requests and plans it creates, along with links to the commits used?

jakethekoenig commented:
> Awesome! Can't wait to see what fine tuned models from this data do.
> 
> Can you post an example or two of the requests and plans it creates, along with links to the commits used?

Oops, I'm glad you asked. When I looked a bit more closely I realized some of the steps had things like "open x.py" and other things that would only be necessary for a human editor. I updated the prompt and think they look better. Some examples:
[60b74](https://github.com/AbanteAI/mentat/commit/60b7458883a52e2ed90f7629fe2743093946ee7b) Prompt: "Add a search command to the system that performs semantic search of files in the code context."
Steps:
1. Modify the CodeContextSettings class to change the attribute name from use_embedding to use_embeddings.
2. In the CodeContext class, replace use_embedding with use_embeddings.
3. Add a search method in CodeContext that takes a query and returns features sorted by similarity score.
4. In commands.py, create a SearchCommand class, and implement the apply method to call the search method in CodeContext.
5. In embeddings.py, change the call_embedding_api function to an async function and add cost and speed tracking.
6. In llm_api.py, update model_context_size and model_price_per_1000_tokens for the ada-002 model.
7. Modify PythonClient and TerminalClient classes to use use_embeddings instead of use_embedding.
8. Update the test_code_context_performance function to use mock_session_context instead of creating a new CodeContextSettings and CodeContext instance.
9. Add a test case in tests/commands_test.py for the search command.

[dd3b3](https://github.com/AbanteAI/mentat/commit/dd3b33d7cf594e012492c514a8f04714178f406d) Prompt: Update diff_context and tests to use absolute file paths for better compatibility and consistency.
Plan:
1. Update the DiffContext class's files property to use absolute paths by obtaining the git_root and appending it to the files in get_files_in_diff.
2. In the tests, remove unnecessary imports of os module and adjust the file paths to use pathlib's Path module to create absolute paths.
3. Update the _update_ops() function's abs_path variable to use pathlib-based paths.
4. Modify the _get_file_message() function to accept an absolute path as an argument
5. Update all calls to the _get_file_message() function, passing the absolute path of the operations.py file.
6. Change all calls to the annotate_file_message() function, passing the absolute path of the operations.py file.

[a00b2](https://github.com/AbanteAI/mentat/commit/a00b22f961cbb14ac31cabfcf09248ae3fde5a77) Prompt: Catch and report shlex errors when processing user commands
Steps:
1. Import the necessary modules for handling errors
2. Modify the collect_user_input function to include error handling
3. Add a try-except block around the shlex.split and command application to catch ValueError exceptions
4. Create a new response to inform the user about the error and provide the error message
